File: python/1260.py
# 1풀이

# import sys
# from collections import deque

# input = sys.stdin.readline

# n, m, v = map(int, input().split())

# s = [[] for _ in range(n+1)]
# s2 = [[] for _ in range(n+1)]

# visited = [False for _ in range(n+1)]
# visited2 = [False for _ in range(n+1)]

# for _ in range(m):
#     a, b = map(int, input().split())
#     s[a][b] = 1
#     s[b][a] = 1
#     s2[a][b] = 1
#     s2[b][a] = 1


# def dfs(v):
#     visited[v] = True
#     print(v, end=' ')
#     for i in range(1, n+1):
#         if s[v][i] == 1 and visited[i] == False:
#             dfs(i)


# def bfs(start):
#     queue = deque([start])
#     visited2[start] = True
#     while(queue):
#         v = queue.popleft()
#         print(v, end=' ')
#         for i in range(1, n+1):
#             if s2[v][i] == 1 and visited2[i] == False:
#                 queue.append(i)
#                 visited2[i] = True

# dfs(v)
# print()
# bfs(v)


# 2 풀이
# 미세하지만 2풀이가 더 효율적
import sys
from collections import deque

input = sys.stdin.readline

n, m, v = map(int, input().split())

s = [[] for _ in range(n+1)]
s2 = [[] for _ in range(n+1)]

visited = [False for _ in range(n+1)]
visited2 = [False for _ in range(n+1)]

for _ in range(m):
    a, b = map(int, input().split())
    s[a].append(b)
    s[b].append(a)
    s2[a].append(b)
    s2[b].append(a)

for i in range(1, len(s)):
    s[i].sort()
    s2[i].sort()


def dfs(v):
    visited[v] = True
    print(v, end=' ')
    for i in s[v]:
        if visited[i] == False:
            dfs(i)


def bfs(start):
    queue = deque([start])
    visited2[start] = True
    while(queue):
        v = queue.popleft()
        print(v, end=' ')
        for i in s[v]:
            if visited2[i] == False:
                queue.append(i)
                visited2[i] = True


dfs(v)
print()
bfs(v)

File: c++/1260.cpp
#include <iostream>
#include <queue>

using namespace std;

int n, m, v;
int map[1001][1001];
bool isVisited[1001] = {false,};
queue<int> q;

void dfs(int v){
  isVisited[v] = true;
  cout << v << " ";
  for(int i=1; i<=n; i++){
    if(map[v][i] && !(isVisited[i])){
      dfs(i);
    }
  }
}

void bfs(int v){
  isVisited[v] = true;
  q.push(v);
  while(!q.empty()){
    v = q.front();
    cout << v << " ";
    q.pop();
    for(int i=1; i<=n; i++){
      if(map[v][i] && !(isVisited[i])){
        q.push(i);
        isVisited[i] = true;
    }
  }
  }
}

int main(){
  cin >> n >> m >> v;

  for(int i=0; i<m; i++){
    int x, y;
    cin >> x >> y;
    map[x][y] = map[y][x] = 1;
  }

  dfs(v);

  for(int i=1; i<=n; i++){
    isVisited[i] = false;
  }
  cout << '\n';

  bfs(v);

  return 0;
}